flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java

```diff
@@ -100,6 +102,16 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
         }
     }
 
+    public void notifyCheckpointSubsumed(long checkpointId) throws Exception {
+        if (!closed) {
+            if (wrapped instanceof AbstractStreamOperator) {
+                ((AbstractStreamOperator<?>) wrapped).notifyCheckpointSubsumed(checkpointId);
+            } else if (wrapped instanceof AbstractStreamOperatorV2) {
+                ((AbstractStreamOperatorV2<?>) wrapped).notifyCheckpointSubsumed(checkpointId);
+            }
```

Comment:
Adding the method was to just follow operator hierarchy and we could extend the usage to other componments ont only keyed state backends if necessary in the future.
If you don't like to add methods in `PublicEvolving` class, I think it's fine and I could revert to call `getKeyedStateBackend()`.

Reply:
That's a tough call. It should follow the hierarchy and `StreamOperator` shouldn't be the public API (we are planning to remove it from the public api at some point). But as it is, it doesn't matter if you add this method to `StreamOperator` or `AbstractStreamOperator`, because all practical cases for implementing a custom `StreamOperator` are always extending from `AbstractStreamOperator`. So I would propose to call `notifyCheckpointSubsumed()` via `getKeyedStateBackend()` for now. Later, once we remove `StreamOperator` from public API we can change it back.

Reply:
Got it, thanks for your explanation.